Playwright: change browser management in BrowserManager.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This patch changes browser instance management occurs which impacts how tests are written.

#### New Files

**Hook file**
This file will contain hooks specifically for the Playwright tests.

**Mocha Configuration**
Mocha configuration file for Playwright test suite is added.

#### Changes to existing files

**TeamCity configuration**
- do not use cached `node_modules` folder when installing dependencies.

**Spec file**
- add multiple steps and subsuites to exercise browser behavior.

**Browser Manager**
- add new functions to manage the three layers of Playwright browser behavior (browser, browser context, page)
- add entrypoint function `start` to be called by test writer.

**Configuration**
- Playwright-specific timeout added.

#### Changes in behavior

BrowserManager is responsible for handling the browser instance, context and pages.
Test writer is expected to call the entrypoint `start` function inside the top-level describe block.

#### Testing

To test, verify the following:
- existing Calypso E2E tests are not affected (i.e. passes).
- Playwright E2E tests should fail with an error in `page.click` as expected.

Partially implements #51082.
Parent: #51419
Child: #51209

worldomonation commented:
Looks like my changes break CircleCI.

worldomonation commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -104,7 +104,7 @@ object RunCalypsoE2eDesktopTests : BuildType({
 				export TEST_FILES=${'$'}(join ',' ${'$'}(find specs*/**/*spec.js -type f -not -path specs-playwright/*))
 
-				yarn magellan --config=magellan.json --max_workers=%E2E_WORKERS% --suiteTag=parallel --local_browser=chrome --mocha_args="--reporter mocha-teamcity-reporter" --test=${'$'}{TEST_FILES}
+				yarn magellan --config=magellan.json --max_workers=%E2E_WORKERS% --suiteTag=parallel --local_browser=chrome --test=${'$'}{TEST_FILES}
```
As mentioned in the top comment, the `mocha_args` flag here was overriding the flags specified in magellan configs. This would have been acceptable when the mocha configuration file loaded `mocha-hooks.js` by default, but this is problematic when mocha config no longer loads a hook file by default.

scinos replied:
We do want to pass `--reporter mocha-teamcity-reporter` here because it is a CI-only thing: if someone runs e2e tests in their local machines, we don't want that reporter to be loaded.

Same for all the other flags really. `magellan.json` should define the defaults for running the tests locally, but in CI we override some of those flags to adapt to the environment.

worldomonation commented on test/e2e/run.sh:
```diff
@@ -104,6 +104,7 @@ while getopts ":a:RpS:B:s:gjCJH:wzyl:cm:f:iIUvxu:h:F" opt; do
       ;;
     g)
       MAGELLAN_CONFIG="magellan.json"
+      MOCHA_ARGS+="-file lib/mocha-hooks/mocha-hooks.js"
```
This is a (hopefully) temporary solution kept only as long as CircleCI tests are run.

noahtallen commented on test/e2e/.mocharc.js:
```diff
@@ -0,0 +1,22 @@
+// existing workflow, if the value is undefined it will default to Selenium.
+if ( framework === undefined || framework.toLowerCase() === 'selenium' ) {
+	configFile = 'lib/mocha-hooks/mocha-hooks.js';
+} else if ( framework.toLowerCase() === 'playwright' ) {
```
Nit: it might make sense to do this instead:

```
const configFile = framework.toLowerCase() === 'playwright' ? //playwright : //selenium
```

Reason being that this would still default to selenium in the case that framework is set to something we don't support

noahtallen commented on test/e2e/lib/types.ts:
```diff
@@ -1,3 +1,6 @@
 export type localeCode = string;
+
+// General
+export type timeoutMS = number;
```
nit: this is simple enough that you really don't have to export it as a custom type -- you could just do `config.get() as number`

worldomonation replied:
Sure, will remove in next commit.

scinos replied:
To be clear, the reason this doesn't help in TypeScript is because it is just an alias. In any place where you expect a `timeoutMS` you can pass a `number` and TS won't complain.

noahtallen commented on test/e2e/magellan.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"mocha_tests": [ "specs" ],
-	"mocha_args": "-R spec-junit-reporter",
+	"mocha_args": "--file lib/mocha-hooks/mocha-hooks.js --reporter mocha-teamcity-reporter -R spec-junit-reporter",
```
Out of curiosity, why doesn't this work via `.mocharc.js`? Or is that used for something else?

worldomonation replied:
That's a good point. I think specifying `.mocharc.js` should work here. I will try it out.

worldomonation replied:
Good call, it works - thanks for pointing this out! 

This also means I should be able to remove my changes to `.teamcity/_self/projects/WebApp.kt` without breaking any of the CircleCI or TeamCity workflows.

noahtallen reviewed (approved):
Looks good, given that the tests pass! I left some minor comments, but nothing blocking

scinos commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -760,6 +760,7 @@ object RunCalypsoPlaywrightE2eTests : BuildType({
 				export LIVEBRANCHES=true
 				export NODE_CONFIG_ENV=test
 				export PLAYWRIGHT_BROWSERS_PATH=0
+				export FRAMEWORK=playwright
```
IMO, the downsides of using an env var are:
- `FRAMEWORK` is too generic. I may have an env var already called `FRAMEWORK` in my laptop that will affect the process when running tests locally
- I may have a `FRAMEWORK=playwright` that I set months ago and forgot about it, which will cause problems if I try to run selenium tests locally.
- I need to specify I want playwright in two separate places and ensure they are in sync (eg: `FRAMEWORK=selenium yarn magellan --config=magellan-playwright.json` is mixing selenium and playwright).

I think it will be simpler if, instead of doing this, we pass `--mocha-args="--config .mocharc-playwright.yml --reporter mocha-teamcity-reporter"` in line 809. And then have two config files, `.mocharc.yml` and `.mocharc-playwright.yml`. Even if there is some duplication between those two files, that will be gone once we migrate all tests (or we move away from Mocha)

worldomonation replied:
I initially had it implemented as such, but I chose to do the unified mocharc.js file with conditional logic because:
- maintain existing workflow for users when they run `./run.sh` or `mocha specs/whatever`
- broke CircleCI runs, requiring me to add  `MOCHA_ARGS+='--config <file>` in one of the flags.

worldomonation replied:
> I need to specify I want playwright in two separate places and ensure they are in sync (eg: FRAMEWORK=selenium yarn magellan --config=magellan-playwright.json is mixing selenium and playwright).

This would be an intentional mix-up but is a real concern that I did not think about.

worldomonation replied:
Since my most recent push broke all tests, I will explore reverting to having two sets of mocha config files. It would make things more explicit and intentional, both good qualities to have for a test suite.

worldomonation committed:
Use yarn install command in the Playwright E2E Test configuration for TeamCity.

worldomonation committed:
Implement mocha hooks for Playwright tests.

The function to close browser upon completion of tests has been moved to the mocha hook.
This ensures the browser is closed at conclusion of each test suite without test writer intervening.

To support this, the following have been changed or added:
- specify the use of a YAML configuration file for mocha specific to Playwright test suite in TeamCity configuration.
- removal of the closeBrowser call from the test.
- creation of a new Playwright hook file and directory under test/e2e/lib.

worldomonation commented:
I've force-pushed a new branch since the older iteration of this PR was quite stale and did not incorporate the changes that were implemented in parent PRs, notably #51419 where BrowserManager was moved to be its own package.

Description will be updated accordingly.

worldomonation commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -733,7 +733,7 @@ object RunCalypsoPlaywrightE2eTests : BuildType({
 
 				# Install modules
-				${_self.yarn_install_cmd}
+				yarn install
```
I've had to change this to explicitly run `yarn install` since the pre-populated `node_modules` cache from [this change](https://github.com/Automattic/wp-calypso/commit/7efc3246ac99b30f24a6040fc4bfeaf3791f7d10#diff-c08774f332e4c025ce1da0c7036888d494b6d981ca4b073a4f2266ec55a6bfbc) does not include Playwright binaries.

noahtallen replied:
Any clue why that's the case? the directory should be a copy of the output from `yarn install` once daily. So it would be weird to me that the binaries aren't there even after a second yarn install, but running `yarn install` by itself does include them :thinking:

worldomonation replied:
> Any clue why that's the case?

I suspect it is because of the environment variable exported above:
`export PLAYWRIGHT_BROWSERS_PATH=0`
without this, it seems the Playwright binaries are not installed in the local `node_modules` directory.

I suppose I can try adding the export to [this file](https://github.com/Automattic/wp-calypso/blob/7efc3246ac99b30f24a6040fc4bfeaf3791f7d10/.teamcity/_self/yarnInstall.kt). My concern was that it would affect non-Playwright jobs.

Do you think that change is safe to make?

noahtallen replied:
Gotcha! In that case the current case makes sense to me. I think if `yarn` has an existing node_modules which is mostly up to date, it must not be trying to do the install step which handles that variable. Thanks for the explanation :+1:

worldomonation replied:
With that said, I should put a reminder somewhere to make that change at some point, so all E2E definitions use the same call. Just reduces possibility that something somewhere goes wrong. So thanks for pointing this out!

worldomonation committed:
Change how configuration files for Magelland and Mocha are specified at each level.

- move the reporter config value to the mocha configuration file (from the TeamCity configuration file).
- remove mocha_args from magellan configuration file.
- add to magellan configuration file the path to mocha configuration file.

worldomonation committed:
Correct issues with the after hook for Playwright.

- remove unnecesary import that was causing failures (after is not a function).
- update the return value to Promise<void> as per Playwright documentation.
- provide a name for the after hook.

worldomonation committed:
Move the before hook from spec file to the hook.

- the before hook to launch a new Browser instance has been moved to the hooks file.

worldomonation committed:
Change the hook for launching browser tab to a before all hook.

worldomonation commented on test/e2e/.mocharc_playwright.yml:
```diff
@@ -0,0 +1,9 @@
+  - 'lib/before.js'
+  - 'lib/hooks-playwright/hooks.ts'
+reporter:
+  - 'mocha-teamcity-reporter'
```
Note that some of the command line arguments under `mocha_args` in TeamCity has been moved here.

scinos replied:
The goal of having some args here and some pass in from TeamCity's shell command is that the arguments here should be the defaults when a developer runs the tests in their machine, and have the TeamCity specific things (like the reporter) only in the shell command.

worldomonation replied:
Restored the command line argument for reporter type in TeamCity configuration in a subsequent pushes.

I've updated this field to be the `spec-junit-reporter`, please let me know if I should not do that.

worldomonation committed:
Specify the reporter type for TeamCity back in the configuration.

worldomonation committed:
Specify 'spec-junit-reporter' as the default reporter type in mocharc.

worldomonation commented:
Looking at the TeamCity runs for both existing Selenium suite as well as Playwright suite, tests appear to run as expected.

noahtallen commented on .teamcity/_self/projects/WebApp.kt:
```diff
@@ -733,7 +733,7 @@ object RunCalypsoPlaywrightE2eTests : BuildType({
 
 				# Install modules
-				${_self.yarn_install_cmd}
+				yarn install
```
Could we add a comment about why we need to go with the raw yarn install command?

WunderBart replied:
Should we be skipping dev & optional deps here? Also, can we use the pure/frozen lockfile option? i.e.

```suggestion
				yarn install --frozen-lockfile --production --ignore-optional
```

cc @scinos

worldomonation replied:
I've incorporated the suggested changes here. Please take a look!

noahtallen replied:
I think we can just keep it with `yarn install`, that's what we use everywhere else: https://github.com/Automattic/wp-calypso/blob/7efc3246ac99b30f24a6040fc4bfeaf3791f7d10/.teamcity/_self/yarnInstall.kt#L16

Also, I think we need to install dev deps (and possibly optional ones too). Wouldn't something like test utils be added to dev dependencies?

noahtallen replied:
Actually, @worldomonation, I think a better way to solve this problem would be to set the `use_cached_node_modules` parameter to false for this build type.

noahtallen replied:
You can just add this after the "description" property in the build type:
```kt
params {
	param("use_cached_node_modules", "false")
}
```

worldomonation replied:
@noahtallen Like this I suppose?

```
	description = "Runs Calypso e2e tests using Playwright"
	params {
	    param("use_cached_node_modules", "false")
}
```

noahtallen replied:
yep, exactly!

noahtallen commented on test/e2e/specs-playwright/wp-log-in-out-spec.js:
```diff
@@ -17,21 +16,15 @@ const mochaTimeOut = config.get( 'mochaTimeoutMS' );
+	// This is akin to the following call in the Selenium test:
+	// driver = await driverManager.startBrowser();
+	before( 'Open new test tab', async function () {
+		page = await this.browserContext.newPage();
```
Out of curiosity, will this have to be added for all tests? If so, does it make sense to abstract it more?

worldomonation replied:
I would like to abstract it some more, definitely. In my previous attempts I ended up running into issues with mocha contexts, but I'll take another crack at it!

worldomonation replied:
Since my last message there was the big slack thread about Mocha + Playwright ([here](p1620084874022800-slack-C01QG4Y91RR)).

I think I was able to come to a reasonable level of abstraction for the time being, where hooks:
- launch the browser
- create a new browser context
- create a new page, which is where the tests execute in

As Sergio noted in the slack thread, in the future it would be a good idea to override the behavior of describe blocks with our own `e2e_describe` that encapsulates the actions listed above. However, that is not within the scope of this PR.

noahtallen replied:
That sounds good to me!

noahtallen reviewed (approved):
Just a few non-blocking comments. Since the tests are passing, this LGTM!

worldomonation committed:
Add default mocha startup timeout to mocha config.

Add browser startup config value when launching a new instance of Browser. This will be used throughout the test by Playwright when interacting with elements on pages.

worldomonation committed:
Update test spec file with multiple subsuites as well as an expected failing suite.

Change hook to launch new browser, then browsercontext then a new page at beginning of test run.

worldomonation committed:
Incorporate changes to TeamCity configuration as per comments in code.

worldomonation committed:
Move @types/config in calypso-e2e package to prod.

worldomonation committed:
Pin the version of @types/config.

worldomonation committed:
Place the @parallel tag for the demonstration tests, otherwise magellan cannot find it.

worldomonation committed:
Revert to using default yarn install command. Instead, flip the parameter to force installation not from the cache.

worldomonation committed:
Set use_cached_node_modules to false.

worldomonation commented:
Changes are again ready to be reviewed. 

Description will be updated shortly.

noahtallen commented on packages/calypso-e2e/package.json:
```diff
@@ -15,12 +15,11 @@
 	],
 	"license": "GPL-2.0-or-later",
 	"dependencies": {
+		"@types/config": "^0.0.38",
```
We should be able to move this back to dev dependencies now that yarn install is fixed

worldomonation replied:
Ok, will do that.

worldomonation replied:
Pushed subsequent commit that reverts this change.

worldomonation committed:
Revert changes to packages/calypso-e2e/package.json.

worldomonation committed:
Turn the `after` hook to an async function.

scinos commented on test/e2e/lib/hooks-playwright/hooks.ts:
```diff
@@ -0,0 +1,30 @@
+ * @returns {Promise<void>} Void promise.
+ */
+after( 'Close browser', async function () {
+	return await BrowserManager.browser.close();
```
Nit: Mocha won't do anything with the returned value of an async hook, so no need to return anything.
```suggestion
	await BrowserManager.browser.close();
```

worldomonation replied:
Got it, in that case I should also update the return value in the JSDoc as well.

worldomonation committed:
Incorporate code review suggestion from Sergio.

worldomonation committed:
Change (yet again) how the browser instances are managed.

This flip-flop is due to knowledge gained from subsequent work I've been doing to implement screenshot and video capture functionality.
It turns out that as mocha hooks do not have in ways I had expected, there is a limit to the level of abstraction I can do using a separate hook file.
For the most reliable test behavior, and to facilitate screenshot and video capture in later PRs, it turns out the best way to launch a browser is to do so inside the top level describe block.

worldomonation commented:
My experiment to fully leverage hooks to handle browser instance management has come to an end, and I have updated the contents of this PR, title and description accordingly.

The issues that I could not resolve adequately are as follows:
1. I had expected the `beforeEach` and `afterEach` hooks to run before and after each `describe` block. This was not the case; instead it runs for literally every `describe` and `step` blocks.
2. using a separate hook file, there was a limit to how granular I can control where each hooks are to be run. For example, having a `afterEach(captureVideo)` hook causes it to run for every block of code, not for each top-level describe block as I had wanted.

In subsequent PRs I had in the pipeline I implemented the screenshot and video capture functionality. Video especially proved to be difficult to work with with the hook-based browser instance management.

For now, I would like to proceed to implementing functionality rather than perfecting the hook-based instance management. This means some more boilerplate for test writers, but a simplified and more reliable browser behavior.

I think that Sergio's idea about replacing the `describe` block with our own block that encapsulates setup and teardown process would be a good investment.

worldomonation committed:
Forgot to include the test spec.

noahtallen reviewed (approved):
I'm happy moving forward with this as a base for future improvements :+1:
